File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/common/widget/dialog/CommonDialog.java
package com.voiceassist.lixinyu.voiceassist.common.widget.dialog;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import com.voiceassist.lixinyu.voiceassist.R;
import com.voiceassist.lixinyu.voiceassist.utils.DisplayUtils;

/**
 * Created by lilidan on 2018/1/25.
 */

public class CommonDialog extends Dialog {

    protected Context context;
    protected int height, width;
    protected boolean cancelTouchout;
    protected View view;



    public CommonDialog(Builder builder) {
        this(builder, R.style.editDialog);
    }

    private CommonDialog(Builder builder, int resStyle) {
        super(builder.context, resStyle);
        context = builder.context;
        height = builder.height;
        width = builder.width;
        cancelTouchout = builder.cancelTouchout;
        view = builder.view;

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(view);

        setCanceledOnTouchOutside(cancelTouchout);


        Window win = getWindow();
        WindowManager.LayoutParams lp = win.getAttributes();
        lp.gravity = Gravity.CENTER;

        if (height > 0) lp.height = height;
        if (width > 0) lp.width = width;

        win.setAttributes(lp);
    }


    public static class Builder {

        protected Context context;
        protected int height, width;
        protected boolean cancelTouchout;
        protected View view;
        protected int resStyle = -1;



        public Builder(Context context) {
            this.context = context;
        }

        public Builder view(int resView) {
            view = LayoutInflater.from(context).inflate(resView, null);
            return this;
        }

        public Builder heightpx(int val) {
            height = val;
            return this;
        }

        public Builder widthpx(int val) {
            width = val;
            return this;
        }

        public Builder heightdp(int val) {
            height = DisplayUtils.INSTANCE.dip2px(context, val);
            return this;
        }

        public Builder widthdp(int val) {
            width = DisplayUtils.INSTANCE.dip2px(context, val);
            return this;
        }

        public Builder heightDimenRes(int dimenRes) {
            height = context.getResources().getDimensionPixelOffset(dimenRes);
            return this;
        }

        public Builder widthDimenRes(int dimenRes) {
            width = context.getResources().getDimensionPixelOffset(dimenRes);
            return this;
        }

        public Builder style(int resStyle) {
            this.resStyle = resStyle;
            return this;
        }

        public Builder cancelTouchout(boolean val) {
            cancelTouchout = val;
            return this;
        }

        public Builder addViewOnclick(int viewRes,View.OnClickListener listener){
            view.findViewById(viewRes).setOnClickListener(listener);
            return this;
        }


        public CommonDialog build() {
            if (resStyle != -1) {
                return new CommonDialog(this, resStyle);
            } else {
                return new CommonDialog(this);
            }
        }
    }
}

File: app/src/main/java/com/voiceassist/lixinyu/voiceassist/common/widget/dialog/CommonDialog.kt
package com.voiceassist.lixinyu.voiceassist.common.widget.dialog

import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.Window
import android.view.WindowManager

import com.voiceassist.lixinyu.voiceassist.R
import com.voiceassist.lixinyu.voiceassist.utils.DisplayUtils

/**
 * Created by lilidan on 2018/1/25.
 */

open class CommonDialog private constructor(builder: Builder, resStyle: Int) : Dialog(builder.context, resStyle) {

    protected var mContext: Context
    protected var height: Int = 0
    protected var width: Int = 0
    protected var cancelTouchout: Boolean = false
    protected var view: View


    constructor(builder: Builder) : this(builder, R.style.editDialog) {}

    init {
        mContext = builder.context
        height = builder.height
        width = builder.width
        cancelTouchout = builder.cancelTouchout
        view = builder.view

    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(view)

        setCanceledOnTouchOutside(cancelTouchout)


        val win = window
        val lp = win!!.attributes
        lp.gravity = Gravity.CENTER

        if (height > 0) lp.height = height
        if (width > 0) lp.width = width

        win.attributes = lp
    }


    open class Builder(var context: Context) {
        var height: Int = 0
        var width: Int = 0
        var cancelTouchout: Boolean = false
        lateinit var view: View
        protected var resStyle = -1

        fun view(resView: Int): Builder {
            view = LayoutInflater.from(context).inflate(resView, null)
            return this
        }

        fun heightpx(`val`: Int): Builder {
            height = `val`
            return this
        }

        fun widthpx(`val`: Int): Builder {
            width = `val`
            return this
        }

        fun heightdp(`val`: Int): Builder {
            height = DisplayUtils.dip2px(context, `val`.toFloat())
            return this
        }

        fun widthdp(`val`: Int): Builder {
            width = DisplayUtils.dip2px(context, `val`.toFloat())
            return this
        }

        fun heightDimenRes(dimenRes: Int): Builder {
            height = context.resources.getDimensionPixelOffset(dimenRes)
            return this
        }

        fun widthDimenRes(dimenRes: Int): Builder {
            width = context.resources.getDimensionPixelOffset(dimenRes)
            return this
        }

        fun style(resStyle: Int): Builder {
            this.resStyle = resStyle
            return this
        }

        fun cancelTouchout(`val`: Boolean): Builder {
            cancelTouchout = `val`
            return this
        }

        fun addViewOnclick(viewRes: Int, listener: View.OnClickListener): Builder {
            view.findViewById<View>(viewRes).setOnClickListener(listener)
            return this
        }


        open fun build(): CommonDialog {
            return if (resStyle != -1) {
                CommonDialog(this, resStyle)
            } else {
                CommonDialog(this)
            }
        }
    }
}
